### DISCUSSION TITLE
Help me separate completions into their own file :pray:

### DISCUSSION THREAD

**Discussion by inyutin** [nushell/nushell#13576] [Q&A]
Hello! 

I'm trying to separate completions (carapace + zoxide) into their own, but I fail to do so.
I create a file (completions.nu)
```nushell
def carapace_completer [] {|spans|
  # if the current command is an alias, get it's expansion
  let expanded_alias = (scope aliases | where name == $spans.0 | get -i 0 | get -i expansion)

  # overwrite
  let spans = (if $expanded_alias != null  {
    # put the first word of the expanded alias first in the span
    $spans | skip 1 | prepend ($expanded_alias | split row " " | take 1)
  } else {
    $spans
  })

  carapace $spans.0 nushell ...$spans
  | from json
}

def zoxide_completer [] {|spans|
    $spans | skip 1 | zoxide query -l ...$in | lines | where {|x| $x != $env.PWD}
}

export def multiple_completions [] {|spans|
    match $spans.0 {
        z => zoxide_completer
        zi => zoxide_completer
        _ => carapace_completer
    } | do $in $spans
}
```

Then in my config.nu I try to do:
```nushell
use "~/.config/nushell/completions.nu" multiple_completions
#....
  completions: {
    case_sensitive: false
    quick: true
    partial: true 
    algorithm: "fuzzy"
    external: {
      enable: true
      max_results: 100
      completer: (multiple_completions)
   }
```

My understanding is that "completer" argument of the config expects to get a closure (an anonymous function) that receives spans as an argument. So in my example my expectation is that  multiple_completions gets called and returns correct closure as expected.

However nu returns error:
```
  × Variable not found
    ╭─[/home/dmitry/.config/nushell/completions.nu:22:11]
 21 │ export def multiple_completions [] {|spans|
 22 │     match $spans.0 {
    ·           ───┬──
    ·              ╰── variable not found
 23 │         z => zoxide_completer
    ╰────
```
I don't understand why it can't find spans if it's an argument right there. What am I missing, where am I wrong ?
I would appreciate help there, thank you

**Comment by inyutin** [comment]
While I'm still wondering why the above doesn't work and I would like to learn why, I've found another solution to separate files

**Comment by ghost** [comment]
maybe there is a weird excepction here, but to me the `def` look weird.  
how you would usually define a function is `def name [ARGMUMENTS] { function_content }` and not `def name [] {|ARGUMENTS| function_content}`.
if i try yours in an interactive terminal it does not complain (maybe it thinks the function returns a closure?), but it also does not seem to work.
```
~> def test [] {|hi| print $hi}
~> test foo
Error: nu::parser::extra_positional

  × Extra positional argument.
   ╭─[entry #2:1:6]
 1 │ test foo
   ·      ─┬─
   ·       ╰── extra positional argument
   ╰────
  help: Usage: test

~> test
Error: nu::shell::variable_not_found

  × Variable not found
   ╭─[entry #1:1:25]
 1 │ def test [] {|hi| print $hi}
   ·                         ─┬─
   ·                          ╰── variable not found
   ╰────
```

have you tried it with a traditional function or traditional closures (`let foo = {|args| print $args}`)?
